libshaderc_spvc/src/spvc_private.cc

```diff
@@ -305,8 +305,12 @@ shaderc_spvc_compilation_result_t generate_hlsl_shader(
       consume_spirv_tools_message, result, std::placeholders::_1,
       std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
 
-  spirv_cross::ParsedIR ir;
-  ir.spirv = std::vector<uint32_t>(source, source + source_len);
+  spirv_cross::ParsedIR* ir = new (std::nothrow) spirv_cross::ParsedIR();
```

Comment:
This ir object leaks now.
This part was better the way you first proposed it.

Note that registering the pass captures the pointer to the ir object, so you have to modify it to say `... spvtools::opt::SpvcIrPass(&ir)...`.    It's ok to have the pass have a pointer to the IR object because the pass will run while the IR object still exists on the stack.